EIPS/eip-6997.md

```diff
@@ -0,0 +1,373 @@
+            super._approve(v.approve, v.tokenId);
+        }
+        else {
+            super._setApprovalForAll(v.owner, v.approve, true);
+        }
+
+        _approvalValidations[approvalId].valid = true;
+    }
+
+    /**
+     * @dev Create a transfer petition of `tokenId` from `from` to `to`.
+     *
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - `tokenId` token must be owned by `from`.
+     *
+     * Emits a {TransferValidate} event.
+     */
+    function _transfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override {
+        require(ERC721.ownerOf(tokenId) == from, "ERC6997: transfer from incorrect owner");
+        require(to != address(0), "ERC6997: transfer to the zero address");
+
+        if(_msgSender() == from) {
+            TransferValidation memory v;
+
+            v.from = from;
+            v.to = to;
+            v.tokenId = tokenId;
+
+            _transferValidations[_totalTransferValidations] = v;
+
+            emit ValidateTransfer(from, to, tokenId, _totalTransferValidations);
+
+            _totalTransferValidations++;
+        } else {
+            super._transfer(from, to, tokenId);
+        }
+    }
+
+    /**
+     * @dev Create an approval petition from `to` to operate on `tokenId`
+     *
+     * Emits an {ValidateApproval} event.
+     */
+    function _approve(address to, uint256 tokenId) internal override virtual {
+        ApprovalValidation memory v;
+
+        v.owner = ownerOf(tokenId);
+        v.approve = to;
+        v.tokenId = tokenId;
+
+        _approvalValidations[_totalApprovalValidations] = v;
+
+        emit ValidateApproval(v.owner, to, tokenId, false, _totalApprovalValidations);
+
+        _totalApprovalValidations++;
+    }
+
+    /**
+     * @dev If approved is true create an approval petition from `operator` to operate on
+     * all of `owner` tokens, if not remove `operator` from operate on all of `owner` tokens
+     *
+     * Emits an {ValidateApproval} event.
+     */
+    function _setApprovalForAll(
+        address owner,
+        address operator,
+        bool approved
+    ) internal override virtual {
+        require(owner != operator, "ERC6997: approve to caller");
+
+        if(approved) {
+            ApprovalValidation memory v;
+
+            v.owner = owner;
+            v.approve = operator;
+            v.approveAll = true;
+
+            _approvalValidations[_totalApprovalValidations] = v;
+
+            emit ValidateApproval(v.owner, operator, 0, true, _totalApprovalValidations);
+
+            _totalApprovalValidations++;
+        }
+        else {
+            super._setApprovalForAll(owner, operator, approved);
+        }
+    }
+}
+```
+
+## Security Considerations
+
+The `_validateTransfer` and `_validateApproval` functions are created as internal, tries to create a secure system by which they can be called.
```

Comment:
Hm, you'll want to focus more on the security considerations of the proposal as a whole (and not just of the reference implementation.) Hypothetically, if you remove the Reference Implementation, the rest of the proposal should still make sense.

Reply:
I understand, I edited the Security Considerations. 👍